File: examples/lua/bittorrent/tracker.lua
-- Copyright (c) 2012, 2014. The SimGrid Team.
-- All rights reserved.

-- This program is free software; you can redistribute it and/or modify it
-- under the terms of the license (GNU LGPL) which comes with this package.

-- A SimGrid Lua implementation of the Bittorrent protocol.

require("simgrid")

common_tracker = {
	MAXIMUM_PEERS = 50
}



function tracker(...)
	tracker_data = {
		peers_list = {},
		deadline = 0,
		comm_received = nil
	}
  -- Check the arguments
  local args = {...}
  if #args ~= 1 then
  	simgrid.info("Wrong number of arguments for the tracker")
  end
  -- Retrieve the end time
  tracker_data.deadline = tonumber(args[1])
  
  simgrid.info("Tracker launched")
  
  local now = simgrid.get_clock()
  
  tracker_data.comm_received = simgrid.task.irecv("tracker")
  while now < tracker_data.deadline do
  	task, err = tracker_data.comm_received:test()
	if task then
  		simgrid.debug("Received a request from " .. task.mailbox)
  		tracker_data.comm_received = simgrid.task.irecv("tracker")
		-- Sending peers to the peer
		local peers = {}
		local i = 0 
		if #tracker_data.peers_list > 0 then
			i = math.random(1,#tracker_data.peers_list)
		end
		while #peers < #tracker_data.peers_list and #peers < common_tracker.MAXIMUM_PEERS do
			table.insert(peers,tracker_data.peers_list[i])
			i = (i % #tracker_data.peers_list) +1
		end
		task.type = "ANSWER"
		task.peers = peers
  		-- Add the peer to our peer list
  		table.insert(tracker_data.peers_list,task.peer_id)
		-- Setting the interval
		task.interval = TRACKER_QUERY_INTERVAL
		-- Sending the task back to the peer
		task:dsend(task.mailbox)
 	else
  		simgrid.process.sleep(1)
  		now = simgrid.get_clock()
  	end
  end
  
  simgrid.info("Tracker is leaving")
end

File: examples/java/bittorrent/Tracker.java
/* Copyright (c) 2006-2014, 2026. The SimGrid Team.
 * All rights reserved.                                                     */

/* This program is free software; you can redistribute it and/or modify it
 * under the terms of the license (GNU LGPL) which comes with this package. */

package bittorrent;
import java.util.ArrayList;

import org.simgrid.msg.Msg;
import org.simgrid.msg.Comm;
import org.simgrid.msg.Host;
import org.simgrid.msg.Task;
import org.simgrid.msg.Process;
import org.simgrid.msg.RngStream;
import org.simgrid.msg.MsgException;

public class Tracker extends Process {
  protected RngStream stream;
  protected ArrayList<Integer> peersList;
  protected double deadline;
  protected Comm commReceived = null;

  public Tracker(Host host, String name, String[]args) {
    super(host,name,args);
  }

  @Override
  public void main(String[] args) throws MsgException {
    if (args.length != 1) {
      Msg.info("Wrong number of arguments for the tracker.");
      return;
    }
    //Build the RngStream object for randomness
    stream = new RngStream("tracker");
    //Retrieve the end time
    deadline = Double.valueOf(args[0]);
    //Building peers array
    peersList = new ArrayList<Integer>();

    Msg.info("Tracker launched.");    
    while (Msg.getClock() < deadline) {
      if (commReceived == null) {
        commReceived = Task.irecv(Common.TRACKER_MAILBOX);
      }
      try {
        if (commReceived.test()) {
          Task task = commReceived.getTask();
          if (task instanceof TrackerTask) {
            TrackerTask tTask = (TrackerTask)task;
            //Sending peers to the peer
            int nbPeers = 0;
            while (nbPeers < Common.MAXIMUM_PEERS && nbPeers < peersList.size()) {
              int nextPeer;
              do {
                nextPeer = stream.randInt(0, peersList.size() - 1);
              } while (tTask.peers.contains(nextPeer));
              tTask.peers.add(peersList.get(nextPeer));
              nbPeers++;
            }
            //Adding the peer to our list
            peersList.add(tTask.peerId);
            tTask.type = TrackerTask.Type.ANSWER;
            //Setting the interval
            tTask.interval = Common.TRACKER_QUERY_INTERVAL;
            //Sending the task back to the peer
            tTask.dsend(tTask.mailbox);
          }
          commReceived = null;
        } else {
          waitFor(1);
        }
      }
      catch (MsgException e) {
        commReceived = null;
      }
    }
    Msg.info("Tracker is leaving");
  }
}
